### PR TITLE
feature: add `server_rewrite_by*`

### PR DESCRIPTION
**Pull request by xiaobiaozhao**
# Why do we need this PR?

> Currently, only **NGX_HTTP_REWRITE_PHASE** has lua modules in `lua-nginx-module` . If you want use lua code in **NGX_HTTP_SERVER_REWRITE_PHASE** phase, you have to develop nginx C modules. So I make this PR solve the problem, you can use this PR's code run lua code in **NGX_HTTP_SERVER_REWRITE_PHASE** phase.

## Add 3 directive
Use like other `*_by_lua*` directives.
```
server_rewrite_by_lua
server_rewrite_by_lua_block
server_rewrite_by_lua_file
```
These three directives work in the **NGX_HTTP_SERVER_REWRITE_PHASE** phase.

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

File: src/ngx_http_lua_module.c [hunk 1]
```diff
@@ -290,6 +291,30 @@ static ngx_command_t ngx_http_lua_cmds[] = {
       (void *) ngx_http_lua_filter_set_by_lua_file },
 #endif
 
+    /* server_rewrite_by_lua "<inline script>" */
+    { ngx_string("server_rewrite_by_lua"),
```
Comment (agentzh) [inline src/ngx_http_lua_module.c:295]: Newer directives should not add this form anymore. Due to the escaping hell.

File: src/ngx_http_lua_module.c [hunk 2]
```diff
@@ -290,6 +291,30 @@ static ngx_command_t ngx_http_lua_cmds[] = {
...
+        NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_CONF_TAKE1,
+        ngx_http_lua_server_rewrite_by_lua,
+        NGX_HTTP_LOC_CONF_OFFSET,
+        0,
+        (void *) ngx_http_lua_server_rewrite_handler_inline },
+
+    /* server_rewrite_by_lua_block { <inline script> } */
+    { ngx_string("server_rewrite_by_lua_block"),
```
Comment (agentzh) [inline src/ngx_http_lua_module.c:303]: No test cases for these new directives?

**Review by agentzh — CHANGES_REQUESTED** [review]

File: t/167-server-rewite.t [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+--- request
+GET /lua
+--- response_body
+OK
+--- error_log
+is_subrequest:true
+is_subrequest:false
+--- no_error_log
+[error]
```
Comment (cppcoffee) [inline t/167-server-rewite.t:169]: need new line.

File: t/167-server-rewite.t [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+
```
Comment (agentzh) [inline t/167-server-rewite.t:170]: You should use our reindex script from the openresty-devel-utils repo to reformat your test files.

File: t/167-server-rewite.t [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+RE
+--- error_log
+uri is /lua
+--- no_error_log
+[error]
+
+
+=== TEST 9: exec
+--- config
+	server_rewrite_by_lua_block {
```
Comment (agentzh) [inline t/167-server-rewite.t:203]: You should add a test case when using both `server_rewrite_by_lua*` and `rewrite_by_lua*`? Both in the same scope or in different scopes.

**Review by agentzh — CHANGES_REQUESTED** [review]

File: t/167-server-rewite.t (hunk 4 above)
Comment (xiaobiaozhao) [reply t/167-server-rewite.t:177]: done

File: t/167-server-rewite.t (hunk 5 above)
Comment (xiaobiaozhao) [reply t/167-server-rewite.t]: TEST 10

File: t/167-server-rewite.t (hunk 3 above)
Comment (xiaobiaozhao) [reply t/167-server-rewite.t]: done

File: src/ngx_http_lua_directive.c [hunk 6]
```diff
@@ -590,6 +591,88 @@ ngx_http_lua_rewrite_by_lua(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
...
+
+    if (llcf->server_rewrite_src.lengths == NULL) {
+        /* no variable found */
+        cache_key = ngx_http_lua_gen_file_cache_key(cf, value[1].data,
+                                                    value[1].len);
+        if (cache_key == NULL) {
+            return NGX_CONF_ERROR;
+        }
+    }
+
```
Comment (zhuizhuhaomeng) [inline src/ngx_http_lua_directive.c:662]: extra blank line

File: src/ngx_http_lua_directive.c [hunk 7]
```diff
@@ -590,6 +591,88 @@ ngx_http_lua_rewrite_by_lua(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
...
+
+    llcf->server_rewrite_src_key = cache_key;
+    llcf->server_rewrite_handler = (ngx_http_handler_pt) cmd->post;
+
+    lmcf = ngx_http_conf_get_module_main_conf(cf, ngx_http_lua_module);
```
Comment (zhuizhuhaomeng) [inline src/ngx_http_lua_directive.c:667]: need to generate the chunkname like other handlers.

File: t/167-server-rewite.t [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    }
+--- request
+GET /lua
+--- response_body
+server_rewrite_by_lua_block in http
+--- error_log
+server_rewrite_by_lua_block in http
+--- no_error_log
+[error]
+--- SKIP
```
Comment (zhuizhuhaomeng) [inline t/167-server-rewite.t:36]:
```suggestion
```

File: t/167-server-rewite.t [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+        }
+    }
+
+--- request
+GET /lua
+--- response_body
+OK
+--- error_log
+is_subrequest:true
+is_subrequest:false
```
Comment (zhuizhuhaomeng) [inline t/167-server-rewite.t:174]:
```suggestion
is_subrequest:false
is_subrequest:true
```

File: t/167-server-rewite.t [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+            ngx.say("OK")
+        }
+    }
+--- request
+GET /lua
+--- response_body
+OK
+--- error_log
+server_rewrite_by_lua_block in http
+rewrite_by_lua_block in location
```
Comment (zhuizhuhaomeng) [inline t/167-server-rewite.t:257]: need to a test case like this https://github.com/openresty/lua-nginx-module/blob/master/t/041-header-filter.t#L820
 to verify the chunkname.

File: src/ngx_http_lua_server_rewriteby.c [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+            return NGX_HTTP_INTERNAL_SERVER_ERROR;
+        }
+
+        cln->handler = ngx_http_lua_request_cleanup_handler;
+        cln->data = ctx;
+        ctx->cleanup = &cln->handler;
+    }
+    /*  }}} */
+
+    ctx->context = NGX_HTTP_LUA_CONTEXT_REWRITE;
```
Comment (zhuizhuhaomeng) [inline src/ngx_http_lua_server_rewriteby.c:268]: I am not sure if we need to define a new phase NGX_HTTP_LUA_CONTEXT_SERVER_REWRITE

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

File: src/ngx_http_lua_directive.c (hunk 6 above)
Comment (xiaobiaozhao) [reply src/ngx_http_lua_directive.c:662]: done

File: src/ngx_http_lua_directive.c (hunk 7 above)
Comment (xiaobiaozhao) [reply src/ngx_http_lua_directive.c:666]: I don't understand what it means, can you explain it in detail?

File: src/ngx_http_lua_server_rewriteby.c (hunk 11 above)
Comment (zhuizhuhaomeng) [reply src/ngx_http_lua_server_rewriteby.c:268]: @agentzh  what do you think?

**Comment by zhuizhuhaomeng** [comment]
@xiaobiaozhao  please rebase to the latest master code

File: t/167-server-rewite.t (hunk 8 above)
Comment (xiaobiaozhao) [reply t/167-server-rewite.t]: done

**Comment by zhuizhuhaomeng** [comment]
ci failed, please have a look at it.

File: src/ngx_http_lua_directive.c (hunk 7 above)
Comment (zhuizhuhaomeng) [reply src/ngx_http_lua_directive.c:666]: like this   https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_directive.c#L559

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

File: src/ngx_http_lua_server_rewriteby.c (hunk 11 above)
Comment (zhuizhuhaomeng) [reply src/ngx_http_lua_server_rewriteby.c:268]: I confirmed with Yichun, need to add NGX_HTTP_LUA_CONTEXT_SERVER_REWRITE.

File: src/ngx_http_lua_module.c [hunk 12]
```diff
@@ -291,6 +292,22 @@ static ngx_command_t ngx_http_lua_cmds[] = {
...
+        ngx_http_lua_server_rewrite_by_lua_block,
+        NGX_HTTP_LOC_CONF_OFFSET,
+        0,
+        (void *) ngx_http_lua_server_rewrite_handler_inline },
+
+    /* server_rewrite_by_lua_file filename; */
+    { ngx_string("server_rewrite_by_lua_file"),
+        NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_CONF_TAKE1,
+        ngx_http_lua_server_rewrite_by_lua,
+        NGX_HTTP_LOC_CONF_OFFSET,
```
Comment (zhuizhuhaomeng) [inline src/ngx_http_lua_module.c:307]:
```suggestion
        NGX_HTTP_SRV_CONF_OFFSET,
```

File: src/ngx_http_lua_module.c [hunk 13]
```diff
@@ -291,6 +292,22 @@ static ngx_command_t ngx_http_lua_cmds[] = {
       (void *) ngx_http_lua_filter_set_by_lua_file },
 #endif
 
+    /* server_rewrite_by_lua_block { <inline script> } */
+    { ngx_string("server_rewrite_by_lua_block"),
+        NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_CONF_BLOCK|NGX_CONF_NOARGS,
+        ngx_http_lua_server_rewrite_by_lua_block,
+        NGX_HTTP_LOC_CONF_OFFSET,
```
Comment (zhuizhuhaomeng) [inline src/ngx_http_lua_module.c:299]:
```suggestion
        NGX_HTTP_SRV_CONF_OFFSET,
```

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

**Pull request merged**